[FIRRTL][LowerToHW]Emit warning on unprocessed remaining annotations

prithayan opened the pull request:
Emit a warning after lowering `FIRRTL` operations to `HW` dialect, if there are `annotations` remaining on the `FIRRTL` operation.
This is required to detect missing passes in FIRRTL dialect to remove the annotations. 
The warnings are disabled by default.


- Use a set to record all the annotations remaining on different operations. 
- Emit a warning for each annotation only once. So if the same annotation is found on different operations, only one warning is emitted.
- Add an option to `LowerFIRRTLToHW` pass to enable the  warnings
  - The option is disabled by defailt
  - Use`emit-annotation-warning=true` to emit the warnings
- Add option `emit-annotation-warning` to `firtool` , which should be passed to `createLowerFIRRTLToHWPass`
- We also need `--mlir-print-op-on-diagnostic=false` otherwise emit warning crashes

seldridge commented on tools/firtool/firtool.cpp:
```diff
@@ -74,6 +74,10 @@ static cl::opt<bool> inliner("inline",
 static cl::opt<bool> lowerToHW("lower-to-hw",
                                cl::desc("run the lower-to-hw pass"));
 
+static cl::opt<bool> enableAnnotationWarning("emit-annotation-warning",
+                             cl::desc("Run the FIRRTL module inliner"),
+                             cl::init(false));
```
The first line is bikeshedding. The second may be this getting copied from earlier. 😄 

```suggestion
static cl::opt<bool> enableAnnotationWarning("warn-on-unprocessed-annotations",
                             cl::desc("Warn about annotations that were not removed by lower-to-hw"),
                             cl::init(false));
```

prithayan replied:
Updated the command line option for both firtool and circt-opt

seldridge commented on test/Conversion/FIRRTLToHW/lower-to-hw-errors.mlir:
```diff
@@ -1,23 +1,43 @@
+  // expected-warning @+1 {{unprocessed annotation}}
+  firrtl.module @foo(in %io_cpu_flush: !firrtl.uint<1>) 
+    attributes { annotations = [{class = "d"}] } {}
+}
```
It would also be good to add a test that checks the "does not print the same annotation multiple times" part.

prithayan replied:
Added tests to check no warnings with same annotation.

seldridge reviewed (commented):
Nice. Looking forward to having this turned on again. 😅 

Some comments throughout.

prithayan commented on test/Conversion/FIRRTLToHW/lower-to-hw-errors.mlir:
```diff
@@ -1,23 +1,43 @@
-// CHECK:   sv.ifdef.procedural "RANDOMIZE_GARBAGE_ASSIGN"  {
-// CHECK-NEXT:   sv.verbatim "`define RANDOMIZE"
-// CHECK-NEXT:  }
-firrtl.circuit "InvalidBundle" {
```
Yes, mistake! Fixed this.

prithayan commented on test/Conversion/FIRRTLToHW/lower-to-hw-errors.mlir:
```diff
@@ -1,23 +1,43 @@
-    return
+  firrtl.module @unprocessedAnnotations(in %clock: !firrtl.clock, in %reset: !firrtl.uint<1>,
+                            in %cond: !firrtl.uint<1>, in %value: !firrtl.uint<2>) {
+    // expected-warning @+1 {{unprocessed annotation}}
```
Added the annotation class to the checks

prithayan commented on test/Conversion/FIRRTLToHW/lower-to-hw.mlir:
```diff
@@ -250,500 +250,4 @@ firrtl.circuit "Simple" {
-
-  // CHECK-LABEL: hw.module @Print
-  firrtl.module @Print(in %clock: !firrtl.clock, in %reset: !firrtl.uint<1>,
-                       in %a: !firrtl.uint<4>, in %b: !firrtl.uint<4>) {
```
Mistake! fixed this.

prithayan commented:
Thanks @seldridge. 
Fixed the accidental changes to lit tests. 
Updated the check for annotation class, and added tests for same annotation class.
Also updated the command line option.

lattner commented:
Wait a second, why does emitting the warning crash, and why is `--mlir-print-op-on-diagnostic=false` required?  This indicates a larger problem that needs to be fixed.  Further, while it is ok temporarily for staging, we shouldn't be going towards a design that requires the command line option.  

Is it possible to fix the core issue that is causing the crash first, then enable the warning unconditionally?

seldridge commented:
Instead of `op.emitError`, a way around this is to use `emitError(op.getLoc(), ...)`. I thought the problem is that the `op.emit*` will try to drop a note and the note will crash.

lattner commented:
Why is this crashing?  Is the IR malformed?  Can we fix that? :-)

prithayan commented:
I'm investigating the crash, it is because we are trying to print an operation, while the IR is being updated. 
This should be fixable.

lattner commented:
Thanks!  If that is what is happening, then it may be simplest to just queue up a bunch of operations that need to be warned about, and emit the warnings at the end of the pass.

prithayan commented:
- The `emitWarn` is crashing when trying to print the `InstanceOp`. Reason for crash:
    1. If we try to print the old `InstanceOp` operation while traversing the operations in a module, the symbol referring to the module is invalid. Hence `getReferencedModule` crashes. 
    2. The traversal is on the new temporary module body, which has moved(`splice`) the old operations,  https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L825
    3. Seems like the original symbol on the operation gets deleted during the `splice`, hence the invalid  temporary IR. 
- To fix the crash, we don't print the operations by using `mlir::emitWarn`, and just print the source location.
- Included another change not related to the crash, in the last commit, 
  - Use `std::mutex` when updating the `StringSet<> alreadyPrinted` in class `CircuitLoweringState`
  - `CircuitLoweringState` object is shared between threads working on modules, hence the lock is required when updating the set. (https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L208)

lattner commented:
Ok, that makes a lot of sense to me, thank you Prithayan.  What is the plan with the flag - now that this doesn't crash, should this be on all the time?

prithayan commented:
Thanks @lattner , quite a few annotations are still remaining on our current run of federation, so it prints a lot of warnings, which might be confusing. We will enable this flag once most of the pending annotations are removed. 
Merging this now.
